### PR TITLE
implement Contravariant instance for ReactClass

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

File: src/React.purs [hunk 1]
```diff
@@ -259,6 +259,18 @@ type ReactSpec props state eff =
   , componentWillUnmount :: ComponentWillUnmount props state eff
   }
 
+cmapSpec
+  :: forall props' props state eff
+   . (props -> props')
+  -> ReactSpec props' state eff
+  -> ReactSpec props Unit eff
+cmapSpec f sp = (spec unit renderFn) { displayName = sp.displayName <> "Mapped" }
```
Comment (paf31) [inline src/React.purs]: Does `displayName` need to change?

Comment (paf31) [inline src/React.purs]: Also, calling `spec` like this loses information if the `Spec` wasn't already built with just `spec`, right? For example, lifecycle methods will be lost.

**Comment by paf31** [comment]
Thanks, but based on the comments above, I'm inclined to recommend using something like contravariant `Yoneda`, where we embed an extra function in the `Spec` type itself, if users need to map over props like this. I'm inclined to think it'll be needed relatively rarely though.

File: src/React.purs (hunk 1 above)
Comment (coot) [inline src/React.purs]: Hmm, in the spec it's indeed better to leave it to the programmer.

Comment (coot) [inline src/React.purs]: It does hide the original spec, but still one can add new ones to the parent spec.

**Comment by coot** [comment]
Is there contravariant yoneda in purescript?  I just couldn't google it, only Haskell's [version]( http://hackage.haskell.org/package/kan-extensions-5.0.1/docs/Data-Functor-Contravariant-Coyoneda.html) pops out. Maybe we actually need kan-extensions library in purescript.

**Comment by paf31** [comment]
No, I don't think so.

**Pull request closed without merging**